addons/web_editor/static/src/js/editor/snippets.options.js

```diff
@@ -3342,31 +3437,289 @@ registry.BackgroundImage = SnippetOptionWidget.extend({
+            const btnContentInnerDiv = document.createElement('div');
+            btnContentInnerDiv.classList.add('o_we_shape');
+            btnContent.appendChild(btnContentInnerDiv);
+
+            const {shape} = btn.dataset;
+            const shapeEl = btnContent.querySelector('.o_we_shape');
+            shapeEl.classList.add(`o_${shape.replace('/', '_')}`);
+            btn.append(btnContent);
+        });
+        return uiFragment;
+    },
+    /**
+     * @override
+     */
+    async _computeWidgetVisibility(widgetName, params) {
+        if (widgetName === 'shape_none_opt') {
+            return false;
         }
         return this._super(...arguments);
     },
+    /**
+     * Flips the shape on its x/y axis.
+     *
+     * @param {boolean} previewMode
+     * @param {'x'|'y'} axis the axis of the shape that should be flipped.
+     */
+    _flipShape(previewMode, axis) {
+        this._handlePreviewState(previewMode, () => {
+            const flip = new Set(this._getShapeData().flip);
+            if (flip.has(axis)) {
+                flip.delete(axis);
+            } else {
+                flip.add(axis);
+            }
+            return {flip: [...flip]};
+        });
+    },
+    /**
+     * Handles everything related to saving state before preview and restoring
+     * it after a preview or locking in the changes when not in preview.
+     *
+     * @param {boolean} previewMode
+     * @param {function} computeShapeData function to compute the new shape data.
+     */
+    _handlePreviewState(previewMode, computeShapeData) {
+        if (previewMode === 'reset') {
+            if (this.prevShape) {
+                this.$target[0].dataset.oeShapeData = this.prevShape;
+            } else {
+                delete this.$target[0].dataset.oeShapeData;
+            }
+        } else {
+            if (previewMode === true) {
+                this.prevShape = this.$target[0].dataset.oeShapeData;
+            }
+            this._markShape(computeShapeData());
+            if (previewMode === false) {
+                this.prevShape = this.$target[0].dataset.oeShapeData;
+            }
+        }
+
+        // Updates/removes the shape container as needed and gives it the
+        // correct background shape
+        const json = this.$target[0].dataset.oeShapeData;
+        const {shape, colors, flip = []} = json ? JSON.parse(json) : {};
+        const target = this.$target[0];
+        let shapeContainer = target.querySelector(':scope > .o_we_shape');
+        if (!shape) {
+            if (shapeContainer) {
+                shapeContainer.remove();
+            }
+            return;
+        }
+        if (!shapeContainer) {
+            shapeContainer = document.createElement('div');
+            target.prepend(shapeContainer);
+            target.style.position = 'relative';
+            shapeContainer.nextElementSibling.style.position = 'relative';
```

Comment:
@sdegueldre Can you explain this to me? I fear this is not needed at all but in any case, this is not robust enough.

Reply:
I honestly don't remember. Removing it for now, I think it was to handle some edge case but the code has changed so much since then that it's entirely possible it's no longer needed.

Reply:
Same as with parallax, since the shape is absolutely positioned it will show on top of non-positioned elements, but as you said this is not very robust. In parallax we do something like

```
.parallax > * {
    position: relative;
}
```

so I will do something similar instead:

```
[data-oe-shape-data] > * {
    position: relative;
}
```

Reply:
Ok, I'll merge something better once shape is merged anyway ;)